File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/ForbidUncheckedTasksUnderRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that forbids unchecked tasks in completed plan documents.
 */
public enum ForbidUncheckedTasksUnderRule implements HarnessCheckRule {
    INSTANCE;
    private static final String CATEGORY = "forbidUncheckedTasksUnder";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final String directory = catNode.get("parameters").get("directory").asText();
        final String uncheckedPattern = catNode.get("parameters").get("uncheckedTaskPattern").asText();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        final Path dirPath = root.resolve(directory);
        if (!HarnessCheckHelper.isSafeDirectory(root, dirPath)) {
            return List.of();
        }
        final Pattern pattern = Pattern.compile(uncheckedPattern);
        return HarnessCheckHelper.safeFileOrWalk(root, dirPath).stream()
                .filter(f -> f.getFileName().toString().endsWith(".md") && !f.getFileName().toString().equals(".gitkeep"))
                .flatMap(file -> {
                    try {
                        final String text = HarnessCheckHelper.readFile(root, file);
                        return pattern.matcher(text).find()
                                ? Stream.of(new Finding(severity, CATEGORY, "completed plan has unchecked tasks: " + root.relativize(file)))
                                : Stream.empty();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/ForbidUncheckedTasksUnderRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.Finding
import ai.harness.gradle.HarnessCheck
import ai.harness.gradle.HarnessPsiResults
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.readText
import kotlin.io.path.relativeTo
import kotlin.io.path.name

/**
 * Rule that forbids unchecked tasks in completed plans.
 */
object ForbidUncheckedTasksUnderRule : HarnessCheckRule {
    override fun applies(manifest: JsonObject): Boolean {
        val category = "forbidUncheckedTasksUnder"
        val catObj = manifest[category]?.jsonObject ?: return false
        val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
        return enabled
    }

    override fun validate(
        manifest: JsonObject,
        root: Path,
        psiResults: HarnessPsiResults?,
    ): Collection<Finding> {
        val category = "forbidUncheckedTasksUnder"
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        val messagesObj = catObj?.get("messages")?.jsonObject
        if (catObj == null || parametersObj == null || messagesObj == null) {
            return emptyList()
        }
        val directory = HarnessCheck.stringFrom(parametersObj, "directory")
        val uncheckedTaskPattern = HarnessCheck.stringFrom(parametersObj, "uncheckedTaskPattern")
        val dirPath = root / directory
        if (!dirPath.isDirectory()) {
            return emptyList()
        }
        val (files, _) = HarnessCheck.walkSafe(root, dirPath)
        /**
         * Pattern compilation may fail if the pattern is invalid; return empty on silent fallback.
         */
        val pattern =
            try {
                uncheckedTaskPattern.toRegex()
            } catch (e: Exception) {
                val skipped = e.localizedMessage
                return emptyList()
            }
        return files
            .filter { file ->
                file.name.endsWith(".md") && pattern.containsMatchIn(file.readText())
            }.map { file ->
                Finding(
                    HarnessCheck.severityOf(manifest, category),
                    category,
                    HarnessCheck.stringFrom(messagesObj, "default").takeIf { message -> message.isNotEmpty() }
                        ?: "completed plan has unchecked tasks: ${file.relativeTo(root)}",
                )
            }
    }
}
